### DISCUSSION TITLE
AsChangeSet: adding a method which can easily check if there is no change to save

### DISCUSSION THREAD

**Discussion by dajoha** [diesel-rs/diesel#3586] [Ideas]
When updating a table with `diesel::update()`, you can easily get this error:

> There are no changes to save. This query cannot be built

Ok, so for a little struct like this:

```rust
#[derive(AsChangeSet)]
struct Person {
    name: Option<String>,
    age: Option<i8>,
}
```

, I just need to add a preliminary guard test, something like this:

```rust
if matches!(person, Person { name: None, age: None }) {
    return;
}
```

But for a bigger struct it can become tedious.

What about adding a method like `has_changes()` to the trait `AsChangeSet`, which would be automatically implemented when auto-deriving? It would return `false` when all the struct fields are `None`; or to be more precise, it would return `false` each time the error "_There are no changes to save. This query cannot be built_" would be triggered.

I am really not aware of the ins and outs of such a question, but it would be really nice, to my user's point of view:)

What do think about it?

A related issue: https://github.com/diesel-rs/diesel/issues/1996

**Comment by weiznich** [comment]
Thanks for filling this feature request. 
I'm happy to read a concrete proposal for this. I would expect that such an proposal explains how exactly `AsChangeset` should be changed, how that will be supported in the derive and how that will be supported in a backward compatible way.

As an intermediate solution: You can always match on the error returned by diesel to detect this case + ignore that error.

**Reply by dajoha** [reply]
Thank you for your reply; I posted this suggestion a bit too fast I must admit:)

This would be a nonsense to add such a method in the trait `AsChangeset`.

I guess it would not be very difficult to implement it in the derive macro, but it would force every people who implemented the trait manually to add this method, with no reason.

It's just totally useless, even more because it can simply be implemented by an external crate, like https://crates.io/crates/is_empty seems to do (not tested).
